clangtidy metrics not same in warnings ng plugin and sonarqube

Hi Team,

We have clangtidy.log file with below **6** warnings:

```
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:53:33: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
                                ^
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:53:34: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
                                 ^
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:53:35: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
                                  ^
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:53:57: warning: 0x0f is a magic number; consider replacing it with a named constant [cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers]
    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
                                                        ^
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:53:72: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
                                                                       ^
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:53:95: warning: 6 is a magic number; consider replacing it with a named constant [cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers]
    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
                                                                                              ^
```
Warnings NG plugin displays number of issues: **6**, But the sonar is only picking **1** warning in sonarqube. 

We understand that because the line number 53 contains multiple duplicate warnings so sonar-cxx will consider them as single warning. **warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]**

But why are the below warnings ignored completely by sonarqube? 
```
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:53:57: warning: 0x0f is a magic number; consider replacing it with a named constant [cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers]
    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
                                                        ^
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:53:95: warning: 6 is a magic number; consider replacing it with a named constant [cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers]
    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
                                                                                              ^
```

Actual sonarqube clangtidy issue displayed in sonarqube: 1
Expected sonarqube metrics : 2 or 3(ignoring the duplicates)

Is it because the warning is associated with multiple rules? **[cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers]**

Please help us resolve the issue.

Thanks in advance!

Regards,
Akilesh D S

Hi @dsakilesh,

thanks for your feedback.

If an issue is not displayed, you should always check first whether the associated rule is [activated in the Quality Profile used](https://github.com/SonarOpenCommunity/sonar-cxx/wiki/Manage-Quality-Profiles).

Duplicates of issues are filtered out by the cxx plugin. This is done by comparing the filename, rule id, line, column and issue text.

> Base64.cpp:53:95: warning: 6 is a magic number; consider replacing it with a named constant [cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers]

Issues always have only one ruleId. That an issue has multiple ruleIds is not supported so far (I have never seen it before). Is this new in Clang-Tidy?

The SQ API unfortunately does not support this, an issue with multiple ids. The question is also what the plugin should do here. Is this one issue or two?

Regards,

Hi @guwirth 

> The SQ API unfortunately does not support this, an issue with multiple ids. The question is also what the plugin should do here. Is this one issue or two?

We expect to pick 2 clangtidy issues/warnings here. (If the rule id for a warning was only one). We will check why there are 2 rule ids for a single warning.

> Duplicates of issues are filtered out by the cxx plugin. This is done by comparing the filename, rule id, line, column and issue text.

Since you have mentioned that filtering happens based on columns also, we expect then sonar to display 4 warnings for the below clangtidy log but it shows only 1 warning. Why?

```
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:53:33: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
                                ^
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:53:34: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
                                 ^
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:53:35: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
                                  ^
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:53:72: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
```
Regards,

Hi @guwirth 

We have upgraded the sonar-cxx plugin to latest version 2.0.2.

We still have got the same issue.

Content of the clang tidy log (4 warnings):

```
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:66:36: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
      output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2)];
                                   ^
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:66:37: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
      output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2)];
                                    ^
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:66:59: warning: 0x0f is a magic number; consider replacing it with a named constant [cppcoreguidelines-avoid-magic-numbers]
      output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2)];
                                                          ^
/var/jenkins/0/workspace/bes-brc3-common_clang-generation/src/src/Util/Base64.cpp:66:59: warning: 0x0f is a magic number; consider replacing it with a named constant [readability-magic-numbers]
```

SonarQube shows only 3 warnings:

![image](https://user-images.githubusercontent.com/58210462/119770539-f2770880-bed9-11eb-9625-8ee05f3f9085.png)

Even though the column number is different for first 2 warnings it seems to consider it as a single warning. Is it because the first 2 warnings have the same error text and rule id? 
**warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]**

Regards

--- a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/clangtidy/CxxClangTidySensor.java
+++ b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/clangtidy/CxxClangTidySensor.java
@@ -23,7 +23,9 @@ import java.io.File;
 import java.nio.charset.StandardCharsets;
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.LinkedList;
 import java.util.List;
+import java.util.Map;
 import java.util.regex.Pattern;
 import org.sonar.api.batch.sensor.SensorDescriptor;
 import org.sonar.api.config.PropertyDefinition;
@@ -43,11 +45,8 @@ public class CxxClangTidySensor extends CxxIssuesReportSensor {
   public static final String REPORT_PATH_KEY = "sonar.cxx.clangtidy.reportPaths";
   public static final String REPORT_ENCODING_DEF = "sonar.cxx.clangtidy.encoding";
   public static final String DEFAULT_ENCODING_DEF = StandardCharsets.UTF_8.name();
-  private static final Logger LOG = Loggers.get(CxxClangTidySensor.class);
 
-  private static final String REGEX
-                                = "(.+|[a-zA-Z]:\\\\.+):([0-9]+):([0-9]+): ([^:]+): (.+)";
-  private static final Pattern PATTERN = Pattern.compile(REGEX);
+  private static final Logger LOG = Loggers.get(CxxClangTidySensor.class);
 
   public static List<PropertyDefinition> properties() {
     return Collections.unmodifiableList(Arrays.asList(
@@ -66,7 +65,7 @@ public class CxxClangTidySensor extends CxxIssuesReportSensor {
         .defaultValue(DEFAULT_ENCODING_DEF)
         .name("Clang-Tidy Report Encoding")
         .description("Defines the encoding to be used to read the files from `sonar.cxx.clangtidy.reportPaths`"
-          + " (default is `UTF-8`).")
+                       + " (default is `UTF-8`).")
         .category("CXX External Analyzers")
         .subCategory("Clang-Tidy")
         .onQualifiers(Qualifiers.PROJECT)
@@ -74,42 +73,17 @@ public class CxxClangTidySensor extends CxxIssuesReportSensor {
     ));
   }
 
-  /**
-   * Extract the rule id from info
-   *
-   * @param info text (with rule id)
-   * @param defaultRuleId rule id to use if info has no or invalid rule id
-   * @return sting array: ruleId, info
-   */
-  protected static String[] splitRuleId(String info, String defaultRuleId) {
-    String ruleId = defaultRuleId;
-
-    if (info.endsWith("]")) { // [ruleId]
-      for (var i = info.length() - 2; i >= 0; i--) {
-        var c = info.charAt(i);
-        if (!(Character.isLetterOrDigit(c) || c == '-' || c == '.' || c == '_')) {
-          if (c == '[') {
-            ruleId = info.substring(i + 1, info.length() - 1);
-            info = info.substring(0, i - 1);
-          }
-          break;
-        }
-      }
-    }
-    return new String[]{ruleId, info};
-  }
-
   @Override
   public void describe(SensorDescriptor descriptor) {
     descriptor
       .name("CXX Clang-Tidy report import")
-      .onlyOnLanguages("cxx","cpp", "c++", "c")
+      .onlyOnLanguages("cxx", "cpp", "c++", "c")
       .createIssuesForRuleRepository(getRuleRepositoryKey())
       .onlyWhenConfiguration(conf -> conf.hasKey(getReportPathsKey()));
   }
 
   @Override
-  protected void processReport(File report)  {
+  protected void processReport(File report) {
     String reportEncoding = context.config().get(REPORT_ENCODING_DEF).orElse(DEFAULT_ENCODING_DEF);
 
     try ( var scanner = new TextScanner(report, reportEncoding)) {
@@ -119,58 +93,25 @@ public class CxxClangTidySensor extends CxxIssuesReportSensor {
       // c:\a\file.cc:5:20: warning: ... conversion from string literal to 'char *' [clang-diagnostic-writable-strings]
       CxxReportIssue currentIssue = null;
       while (scanner.hasNextLine()) {
-        String nextLine = scanner.nextLine();
-        var matcher = PATTERN.matcher(nextLine);
-        if (matcher.matches()) {
-          // group: 1      2      3         4        5
-          //      <path>:<line>:<column>: <level>: <info>
-          var m = matcher.toMatchResult();
-          String path = m.group(1); // relative paths
-          String line = m.group(2);
-          String column = m.group(3);
-          String level = m.group(4); // error, warning, note, ...
-          String info = m.group(5); // info [ruleId]
-
-          CxxReportIssue newIssue = null;
-          var saveIssue = true;
-
-          switch (level) {
-            case "note":
-              saveIssue = false;
-              if (currentIssue != null) {
-                currentIssue.addFlowElement(path, line, column, info);
-              }
-              break;
-            case "warning": {
-              String [] rule = splitRuleId(info, "clang-diagnostic-warning");
-              newIssue = new CxxReportIssue(rule[0], path, line, column, rule[1]);
+        var issue = Issue.create(scanner.nextLine());
+        if (issue != null) {
+          if ("note".equals(issue.level)) {
+            if (currentIssue != null) {
+              currentIssue.addFlowElement(issue.path, issue.line, issue.column, issue.info);
             }
-              break;
-            case "error":
-            case "fatal error": {
-              String [] rule = splitRuleId(info, "clang-diagnostic-error");
-              newIssue = new CxxReportIssue(rule[0], path, line, column, rule[1]);
-            }
-              break;
-            default: {
-              String [] rule = splitRuleId(info, "clang-diagnostic-unknown");
-              newIssue = new CxxReportIssue(rule[0], path, line, column, rule[1]);
-            }
-              break;
-          }
-          if (saveIssue) {
+          } else {
             if (currentIssue != null) {
               saveUniqueViolation(currentIssue);
-              currentIssue = null;
             }
-            currentIssue = newIssue;
-            newIssue = null;
+            currentIssue = new CxxReportIssue(issue.ruleId, issue.path, issue.line, issue.column, issue.info);
+            for (var aliasRuleId : issue.aliasRuleIds) {
+              currentIssue.addAliasRuleId(aliasRuleId);
+            }
           }
         }
       }
       if (currentIssue != null) {
         saveUniqueViolation(currentIssue);
-        currentIssue = null;
       }
     } catch (final java.io.IOException
                      | java.lang.IllegalArgumentException
@@ -189,4 +130,74 @@ public class CxxClangTidySensor extends CxxIssuesReportSensor {
   protected String getRuleRepositoryKey() {
     return CxxClangTidyRuleRepository.KEY;
   }
+
+  private static class Issue {
+
+    private static final String REGEX = "(.+|[a-zA-Z]:\\\\.+):([0-9]+):([0-9]+): ([^:]+): (.+)";
+    private static final Pattern PATTERN = Pattern.compile(REGEX);
+
+    private String path;
+    private String line;
+    private String column;
+    private String level;
+    private String ruleId;
+    private LinkedList<String> aliasRuleIds = new LinkedList<>();
+    private String info;
+
+    static Issue create(String data) {
+      var matcher = PATTERN.matcher(data);
+      if (matcher.matches()) {
+        var issue = new Issue();
+        // group: 1      2      3         4        5
+        //      <path>:<line>:<column>: <level>: <info>
+        var m = matcher.toMatchResult();
+        issue.path = m.group(1); // relative paths
+        issue.line = m.group(2);
+        issue.column = m.group(3);
+        issue.level = m.group(4); // error, warning, note, ...
+        issue.info = m.group(5); // info [ruleIds]
+
+        issue.splitRuleId();
+        return issue;
+      }
+      return null;
+    }
+
+    void splitRuleId() {
+      ruleId = getDefaultRuleId();
+
+      if (info.endsWith("]")) { // [ruleIds]
+        var end = info.length() - 1;
+        for (var start = info.length() - 2; start >= 0; start--) {
+          var c = info.charAt(start);
+          if (!(Character.isLetterOrDigit(c) || c == '-' || c == '.' || c == '_')) {
+            if (c == ',') {
+              var aliasId = info.substring(start + 1, end);
+              if (!"warnings-as-errors".equals(aliasId)) {
+                aliasRuleIds.addFirst(aliasId);
+              }
+              end = start;
+              continue;
+            } else if (c == '[') {
+              ruleId = info.substring(start + 1, end);
+              info = info.substring(0, start - 1);
+            }
+            break;
+          }
+        }
+      }
+    }
+
+    String getDefaultRuleId() {
+      Map<String, String> map = Map.of(
+        "note", "",
+        "warning", "clang-diagnostic-warning",
+        "error", "clang-diagnostic-error",
+        "fatal error", "clang-diagnostic-error"
+      );
+
+      return map.getOrDefault(level, "clang-diagnostic-unknown");
+    }
+  }
+
 }

--- a/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
+++ b/cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxIssuesReportSensor.java
@@ -63,6 +63,15 @@ public abstract class CxxIssuesReportSensor extends CxxReportSensor {
     }
   }
 
+  private void saveIssue(String ruleId, CxxReportIssue issue) {
+    var newIssue = context.newIssue();
+    if (addLocations(newIssue, ruleId, issue)) {
+      addFlow(newIssue, issue);
+      newIssue.save();
+      savedNewIssues++;
+    }
+  }
+
   /**
    * Saves code violation only if it wasn't already saved
    *
@@ -75,11 +84,12 @@ public abstract class CxxIssuesReportSensor extends CxxReportSensor {
   public void saveUniqueViolation(CxxReportIssue issue) {
     if (uniqueIssues.add(issue)) {
       try {
-        var newIssue = context.newIssue();
-        if (addLocations(newIssue, issue)) {
-          addFlow(newIssue, issue);
-          newIssue.save();
-          savedNewIssues++;
+        saveIssue(issue.getRuleId(), issue);
+        if (issue.hasAliasRuleIds()) {
+          // in case of alias rule ids save the issues also with these ids
+          for (var aliasRuleId : issue.getAliasRuleIds()) {
+            saveIssue(aliasRuleId, issue);
+          }
         }
       } catch (RuntimeException e) {
         var msg = "Cannot save the issue '" + issue + "'";
@@ -139,7 +149,7 @@ public abstract class CxxIssuesReportSensor extends CxxReportSensor {
     return null;
   }
 
-  private boolean addLocations(NewIssue newIssue, CxxReportIssue issue) {
+  private boolean addLocations(NewIssue newIssue, String ruleId, CxxReportIssue issue) {
     var first = true;
     for (var location : issue.getLocations()) {
       NewIssueLocation newLocation = null;
@@ -154,7 +164,7 @@ public abstract class CxxIssuesReportSensor extends CxxReportSensor {
       if (newLocation != null) {
         if (first) {
           newIssue
-            .forRule(RuleKey.of(getRuleRepositoryKey(), issue.getRuleId()))
+            .forRule(RuleKey.of(getRuleRepositoryKey(), ruleId))
             .at(newLocation);
           first = false;
         } else {

--- a/cxx-sensors/src/test/java/org/sonar/cxx/sensors/clangtidy/CxxClangTidySensorTest.java
+++ b/cxx-sensors/src/test/java/org/sonar/cxx/sensors/clangtidy/CxxClangTidySensorTest.java
@@ -89,6 +89,69 @@ public class CxxClangTidySensorTest {
   }
 
   @Test
+  public void shouldReportSameIssueInSameLineWithDifferentColumn() {
+    var context = SensorContextTester.create(fs.baseDir());
+    settings.setProperty(
+      CxxClangTidySensor.REPORT_PATH_KEY,
+      "clang-tidy-reports/cpd.report-cols.txt"
+    );
+    context.setSettings(settings);
+
+    context.fileSystem().add(TestInputFileBuilder
+      .create("ProjectKey", "sources/utils/code_chunks.cpp")
+      .setLanguage("cxx")
+      .initMetadata(
+        "asd\n"
+          + "    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];\n"
+          + "asda\n")
+      .build()
+    );
+
+    var sensor = new CxxClangTidySensor();
+    sensor.execute(context);
+
+    assertThat(context.allIssues()).hasSize(4);
+    var issuesList = new ArrayList<Issue>(context.allIssues());
+    assertThat(issuesList.get(0).ruleKey().rule()).isEqualTo("hicpp-signed-bitwise");
+    assertThat(issuesList.get(0).primaryLocation().textRange().start().lineOffset()).isEqualTo(33);
+    assertThat(issuesList.get(1).ruleKey().rule()).isEqualTo("hicpp-signed-bitwise");
+    assertThat(issuesList.get(1).primaryLocation().textRange().start().lineOffset()).isEqualTo(34);
+    assertThat(issuesList.get(2).ruleKey().rule()).isEqualTo("hicpp-signed-bitwise");
+    assertThat(issuesList.get(2).primaryLocation().textRange().start().lineOffset()).isEqualTo(35);
+    assertThat(issuesList.get(3).ruleKey().rule()).isEqualTo("hicpp-signed-bitwise");
+    assertThat(issuesList.get(3).primaryLocation().textRange().start().lineOffset()).isEqualTo(72);
+  }
+
+  @Test
+  public void shouldReportAliasRuleIds() {
+    var context = SensorContextTester.create(fs.baseDir());
+    settings.setProperty(
+      CxxClangTidySensor.REPORT_PATH_KEY,
+      "clang-tidy-reports/cpd.report-alias-rule-ids.txt"
+    );
+    context.setSettings(settings);
+
+    context.fileSystem().add(TestInputFileBuilder
+      .create("ProjectKey", "sources/utils/code_chunks.cpp")
+      .setLanguage("cxx")
+      .initMetadata(
+        "asd\n"
+          + "    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];\n"
+          + "asda\n")
+      .build()
+    );
+
+    var sensor = new CxxClangTidySensor();
+    sensor.execute(context);
+
+    assertThat(context.allIssues()).hasSize(3);
+    var issuesList = new ArrayList<Issue>(context.allIssues());
+    assertThat(issuesList.get(0).ruleKey().rule()).isEqualTo("cppcoreguidelines-avoid-magic-numbers");
+    assertThat(issuesList.get(1).ruleKey().rule()).isEqualTo("readability-magic-numbers");
+    assertThat(issuesList.get(2).ruleKey().rule()).isEqualTo("test");
+  }
+
+  @Test
   public void shouldReportErrors() {
     var context = SensorContextTester.create(fs.baseDir());
     settings.setProperty(

--- a/cxx-sensors/src/test/resources/org/sonar/cxx/sensors/reports-project/clang-tidy-reports/cpd.report-alias-rule-ids.txt
+++ b/cxx-sensors/src/test/resources/org/sonar/cxx/sensors/reports-project/clang-tidy-reports/cpd.report-alias-rule-ids.txt
@@ -1,0 +1,4 @@
+sources\utils\code_chunks.cpp:2:57: warning: 0x0f is a magic number; consider replacing it with a named constant [cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers,test]
+    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
+                                                        ^
+

--- a/cxx-sensors/src/test/resources/org/sonar/cxx/sensors/reports-project/clang-tidy-reports/cpd.report-cols.txt
+++ b/cxx-sensors/src/test/resources/org/sonar/cxx/sensors/reports-project/clang-tidy-reports/cpd.report-cols.txt
@@ -1,0 +1,12 @@
+sources\utils\code_chunks.cpp:2:33: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
+    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
+                                ^
+sources\utils\code_chunks.cpp:2:34: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
+    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
+                                 ^
+sources\utils\code_chunks.cpp:2:35: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
+    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
+                                  ^
+sources\utils\code_chunks.cpp:2:72: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
+    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
+                                                                       ^

--- a/cxx-squid/src/main/java/org/sonar/cxx/utils/CxxReportIssue.java
+++ b/cxx-squid/src/main/java/org/sonar/cxx/utils/CxxReportIssue.java
@@ -32,15 +32,13 @@ import javax.annotation.Nullable;
 public class CxxReportIssue {
 
   private final String ruleId;
-  private final List<CxxReportLocation> locations;
-  private final List<CxxReportLocation> flow;
+  private final List<String> aliasRuleIds = new ArrayList<>();
+  private final List<CxxReportLocation> locations = new ArrayList<>();
+  private final List<CxxReportLocation> flow = new ArrayList<>();
 
   public CxxReportIssue(String ruleId, @Nullable String file, @Nullable String line, @Nullable String column,
                         String info) {
-    super();
     this.ruleId = ruleId;
-    this.locations = new ArrayList<>();
-    this.flow = new ArrayList<>();
     addLocation(file, line, column, info);
   }
 
@@ -52,8 +50,20 @@ public class CxxReportIssue {
     flow.add(0, new CxxReportLocation(file, line, column, info));
   }
 
+  public final void addAliasRuleId(String aliasRuleId) {
+    aliasRuleIds.add(aliasRuleId);
+  }
+
   public String getRuleId() {
     return ruleId;
+  }
+
+  public boolean hasAliasRuleIds() {
+    return !aliasRuleIds.isEmpty();
+  }
+
+  public List<String> getAliasRuleIds() {
+    return Collections.unmodifiableList(aliasRuleIds);
   }
 
   public List<CxxReportLocation> getLocations() {
@@ -66,14 +76,18 @@ public class CxxReportIssue {
 
   @Override
   public String toString() {
+    String aliasIds = aliasRuleIds.stream().map(Object::toString).collect(Collectors.joining(", "));
     String locationsToString = locations.stream().map(Object::toString).collect(Collectors.joining(", "));
     String flowToString = flow.stream().map(Object::toString).collect(Collectors.joining(", "));
-    return "CxxReportIssue [ruleId=" + ruleId + ", locations=" + locationsToString + ", flow=" + flowToString + "]";
+    return "CxxReportIssue [ruleId=" + ruleId
+             + ", aliasIds=" + aliasIds
+             + ", locations=" + locationsToString
+             + ", flow=" + flowToString + "]";
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(ruleId, locations, flow);
+    return Objects.hash(ruleId, aliasRuleIds, locations, flow);
   }
 
   @Override
@@ -89,6 +103,7 @@ public class CxxReportIssue {
     }
     var other = (CxxReportIssue) obj;
     return Objects.equals(ruleId, other.ruleId)
+             && Objects.equals(aliasRuleIds, other.aliasRuleIds)
              && Objects.equals(locations, other.locations)
              && Objects.equals(flow, other.flow);
   }

--- a/integration-tests/features/clangtidy.feature
+++ b/integration-tests/features/clangtidy.feature
@@ -1,0 +1,19 @@
+@SqApi79
+Feature: Importing Clang-Tidy reports
+  As a CXX plug-in user, I want to import Clang-Tidy reports into SonarQube
+
+  Scenario Outline: Importing Clang-Tidy reports
+    Given the project "clangtidy_project"
+    And rule "clangtidy:hicpp-signed-bitwise" is activated
+    When I run sonar-scanner with "-X -Dsonar.cxx.clangtidy.reportPaths=<reportpaths>"
+    Then the analysis finishes successfully
+    And the analysis in server has completed
+    And the server log (if locatable) contains no error/warning messages
+    And the analysis log contains no error/warning messages except those matching:
+      """
+      .*WARN.*Unable to get a valid mac address, will use a dummy address
+      """
+    And the number of violations fed is <violations>
+    Examples:
+      | reportpaths | violations |
+      | column.txt  | 4          |

--- a/integration-tests/testdata/clangtidy_project/column.txt
+++ b/integration-tests/testdata/clangtidy_project/column.txt
@@ -1,0 +1,12 @@
+src\module1.cc:9:33: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
+    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
+                                ^
+src\module1.cc:9:34: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
+    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
+                                 ^
+src\module1.cc:9:35: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
+    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
+                                  ^
+src\module1.cc:9:72: warning: use of a signed integer operand with a binary bitwise operator [hicpp-signed-bitwise]
+    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
+                                                                       ^

--- a/integration-tests/testdata/clangtidy_project/sonar-project.properties
+++ b/integration-tests/testdata/clangtidy_project/sonar-project.properties
@@ -1,0 +1,20 @@
+# metadata
+sonar.projectKey=clangtidy_project
+
+# disable SCM support
+sonar.scm.disabled=true
+
+# disable XML sensor
+sonar.xml.file.suffixes=.disable-xml
+
+# file extensions assigned to the cxx programming language
+sonar.cxx.file.suffixes=.cxx,.cpp,.cc,.c,.hxx,.hpp,.hh,.h
+
+# comma-separated paths to directories containing source files
+sonar.sources=src
+
+# comma-separated list of directories where the plugin should search for include files
+sonar.cxx.includeDirectories=src
+
+# paths to the cppcheck report
+sonar.cxx.clangtidy.reportPaths=*.txt

--- a/integration-tests/testdata/clangtidy_project/src/module1.cc
+++ b/integration-tests/testdata/clangtidy_project/src/module1.cc
@@ -1,0 +1,10 @@
+void func1()
+{
+    int input[256]{};
+    int table[256]{};
+    int output[256]{};
+    int inputPos{};
+    int outputPos{};
+
+    output[outputPos++] = table[((input[inputPos + 1] & 0x0f) << 2) | (input[inputPos + 2] >> 6)];
+}

